param_passing/tutorial.md

```diff
@@ -0,0 +1,561 @@
+```
+
+In the above example, since `//material` does not have listed children elements,
+the entire `//material` element is removed.
+To remove only individual children of `//material`, list them as follows:
+
+```xml
+<!-- //experimental:params -->
+<visual element_id="chassis::camera_visual">
+  <material action="remove">
+    <diffuse/>
+    <ambient/>
+  </material>
+</visual>
+```
+
+This removes, `//diffuse` and `//ambient` from `//material` but not `//material`.
+
+Note that it is not possible to remove further sub elements.
+For example, if the original model has:
+
+```xml
+<!-- included model -->
+<joint name="joint" type="revolute">
+  ...
+  <axis>
+    ...
+    <dynamics>
+      ...
+      <damping>0.2</damping>
+    </dynamics>
+  </axis>
+</joint>
+```
+
+`//damping` can not be removed using the `remove` action, only `//dynamics` can be.
+Use the `replace` action to replace `//dynamics` with the desired instead
+(see `replace` action example 2).
+
+---
+
+### Replace action
+
+The `replace` action replaces elements from the original model to the new provided elements.
+
+Elements can only be replaced with the same type of elements
+(e.g., a `//link` can not be replaced by a `//visual`).
+When using the `replace` action where `@element_id` is declared,
+then the element requires a `@name` attribute (even if the old name is being used)
+since we are replacing the original element with the newly provided.
+Similar to `remove`, if `replace` is stated in the direct children of the
+`@element_id` element then the `@name` attribute is used to find the child of `@element_id`.
+If none is provided, the first matching element (with the same tag) is assumed to be replaced.
+Replacing `@name` with a new `@name` can only be done when the `replace` action
+is defined in the `@element_id` element.
+
+#### Replace examples
+
+**Example 1**
+
+Given the original model has a `//link[@name="chassis"]/visual[@name="camera_visual"]`,
+to `replace` the `//visual[@name="camera_visual"]` element:
+
+```xml
+<!-- //experimental:params -->
+<visual element_id="chassis::camera_visual" name="camera_visual" action="replace">
+  ...
+</visual>
+```
+
+This replaces the original element with the provided element above
+but notice the `@name` attribute will remain the same.
+
+The following example is equivalent but is not recommended:
+
+```xml
+<!-- //experimental:params -->
+<link element_id="chassis">
+  <visual name="camera_visual" action="replace">
+    ...
+  </visual>
+</link>
+```
+
+In this situation, stating `@name` is not required but is encouraged
+since there could exist multiple `//visual` elements in the original.
+When there are multiple, the first found `//visual` element found is assumed to be replaced.
+Also, in this case, the `@name` can not be changed.
+To `replace` the element with a new name:
+
+```xml
+<!-- //experimental:params -->
+<visual element_id="chassis::camera_visual" name="new_name" action="replace">
+  ...
+</visual>
```

Comment:
Just want to confirm about the following use case. First I change the an element name to a new value. Then a few lines later I want to remove that element:

```
<visual element_id="chassis::camera_visual" name="new_name" action="replace">
  ...
</visual>

...

<visual element_id="chassis::new_name" action="remove"/>
```

The above won't work because the `element_id` in the `remove` action should still be `chassis::camera_visual` and not the new name correct?

Reply:
No, the above would work. Parameter passing works off the original model using a top-down approach (not a copy). So first the `//visual` element will be replaced with `//visual[@name="new_name"]` and then when it reaches `<visual element_id="chassis::new_name" action="remove"/>` it will be removed

Reply:
ok got it. Maybe add a small note mentioning that subsequent actions will need to refer to a new `element_id` after this name change?

Reply:
How's this? fbdc63d2c8d3b17353ac085c2355eecb5f205329

Reply:
looks good, thanks!